File: Problem-2 Valid parentheses.py
class Solution:
    def isValid(self, s) :
        if s == None or len(s) == 0 or len(s) % 2 != 0:
            return False
        stack = []
        n = len(s)
        for i in range(n):
            c = s[i]
            if c == '(':
                stack.append(')')
            elif c == '[':
                stack.append(']')
            elif c == '{':
                stack.append('}')
            else:
                if not stack or c != stack.pop():
                    return False

        if len(stack) == 0:
            return True
        else:
            return False

File: Problem2 Valid Parentheses.java
O(N) time ans space complexity where N is length of the string

class Solution {
    
    public boolean isValid(String s) {
        if (s.length()%2 !=0 ) return false;
        Stack<Character> stack = new Stack();
        for (Character c: s.toCharArray()){
            if(c == '(' || c == '{' || c=='['){stack.push(c);}
            else if(c == ')' &&  !stack.isEmpty() && stack.peek() == '(')
            {stack.pop();}
            else if(c == '}' &&  !stack.isEmpty() && stack.peek() == '{')
            {stack.pop();}
            else if(c == ']' &&  !stack.isEmpty() && stack.peek() == '[')
            {stack.pop();} else return false;
        }
        return stack.isEmpty();
      }
    }
